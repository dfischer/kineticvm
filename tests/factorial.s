	5 -> factorial_in
	%PC -> factorial_ret
	:factorial -> %PC
	factorial_out -> %OUT

	6 -> factorial_in
	%PC -> factorial_ret
	:factorial -> %PC
	factorial_out -> %OUT

	4 -> factorial_in
	%PC -> factorial_ret
	:factorial -> %PC
	factorial_out -> %OUT

	1 -> %HALT

factorial:
	1 -> factorial_out

loop_start:
	; factorial_out += factorial_in
	factorial_in -> %LHS
	factorial_out -> %RHS
	%PRODUCT -> factorial_out

	; factorial_in -= 1
	factorial_in -> %LHS
	-1 -> %RHS
	%SUM -> factorial_in

	; if factorial_in != 0, restart
	factorial_in -> %LHS
	%NOT -> %SKIP_INSTR
	:loop_start -> %PC

	; return
	factorial_ret -> %LHS
	2 -> %RHS
	%SUM -> %PC