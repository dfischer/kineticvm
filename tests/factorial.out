120
720
24
